solr/core/src/java/org/apache/solr/util/circuitbreaker/LoadAverageCircuitBreaker.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.util.circuitbreaker;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.management.ManagementFactory;
+import java.lang.management.OperatingSystemMXBean;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>
+ * Tracks current system load average and triggers if the specified threshold is breached.
+ *
+ * This circuit breaker gets the load average (length of the run queue) over the last
+ * minute and uses that data to take a decision. We depend on OperatingSystemMXBean which does
+ * not allow a configurable interval of collection of data.
+ * //TODO: Use Codahale Meter to calculate the value locally.
+ * </p>
+ *
+ * <p>
+ * The configuration to define which mode to use and the trigger threshold are defined in
+ * solrconfig.xml
+ * </p>
+ */
+public class LoadAverageCircuitBreaker extends CircuitBreaker {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final OperatingSystemMXBean operatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean();
+
+  private final boolean enabled;
+  private final double loadAverageThreshold;
+
+  // Assumption -- the value of these parameters will be set correctly before invoking getDebugInfo()
+  private static final ThreadLocal<Double> seenLoadAverage = ThreadLocal.withInitial(() -> 0.0);
+
+  private static final ThreadLocal<Double> allowedLoadAverage = ThreadLocal.withInitial(() -> 0.0);
+
+  public LoadAverageCircuitBreaker(CircuitBreakerConfig config) {
+    super(config);
+
+    this.enabled = config.getLoadAverageCBEnabled();
+    this.loadAverageThreshold = config.getLoadAverageCBThreshold();
+  }
+
+  @Override
+  public boolean isTripped() {
+    if (!isEnabled()) {
+      return false;
+    }
+
+    if (!enabled) {
+      return false;
+    }
+
+    double localAllowedLoadAverage = getLoadAverageThreshold();
+    double localSeenLoadAverage = calculateLiveLoadAverage();
+
+    if (localSeenLoadAverage < 0) {
+      if (log.isWarnEnabled()) {
```

Comment:
There is a fair bit of overlap between the two circuit breakers -- can we move the common code and methods to a super class, and define the contract which both (and future) CPU circuit breakers need to adhere to?

Reply:
An extra base class only for CPU actually makes the code base more complicated, not less. The other circuit breaker is a memory CB, so this cannot be shared with those.

I can't think of another CPU breaker, so let's follow YAGNI and avoid introducing extra base classes "just in case".

Some quick line counts show that the two files have 115 and 139 lines. A quick diff | wc shows 86 differences between the two files. So there isn't as much overlap as it might appear.

Reply:
I respect fully disagree.

The files have common methods names - which make perfect candidates for a common interface, for eg.

Also, the tests are almost a copy paste of the existing.

This is not a just in case - two concrete circuit breakers share significant code. This can be improved, so should be

Reply:
We disagree about coding style.

This fixes a bug. The existing CPU circuit breaker really is not usable because the inputs are limited to the wrong values.

Reply:
I don't understand your comment.

 Regardless of how we interpret the intention of the two circuit breakers, the fact remains that they have significant API and code overlap, which goes against any coding guideline

Reply:
We could also defer further refactoring to a followup JIRA to split things up a bit. There may be other larger refactorings for 9.0 so perhaps there are synergies?

Reply:
After #1725 this is less of a problem. Users choose one or more CB plugins to activate, and each is named and documented precisely. The legacy `CircuitBreakerManager` is a deprecated plugin of its own, and its `cpuThreshold` has not changed semantics, still uses load avg.